src/v/kafka/server/handlers/fetch.cc

```diff
@@ -671,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+     *
+     * It then waits until the shard_fetch_worker completes
+     * and then notifies the fetch coordinator.
+     */
+    ss::future<> start_shard_fetch_worker(
+      ss::shard_id shard,
+      op_context& octx,
+      shard_fetch fetch,
+      size_t min_fetch_bytes) {
+        // if over budget skip the fetch.
+        if (octx.bytes_left <= 0) {
+            co_return;
+        }
+        // no requests for this shard, do nothing
+        if (fetch.empty()) {
+            co_return;
+        }
+
+        const bool foreign_read = shard != ss::this_shard_id();
+
+        auto results = co_await octx.rctx.partition_manager().invoke_on(
+          shard,
+          octx.ssg,
+          [this,
+           min_fetch_bytes,
+           foreign_read,
+           configs = fetch.requests,
+           &octx](cluster::partition_manager& mgr) mutable {
+              // &octx is captured only to immediately use its accessors here
+              // so that there is a list of all objects accessed next to
+              // `invoke_on`. This is meant to help avoiding unintended cross
+              // shard access
+              return shard_fetch_worker(shard_local_fetch_context{
+                .foreign_read = foreign_read,
+                .bytes_left = octx.bytes_left,
+                .min_bytes = min_fetch_bytes,
+                .deadline = octx.deadline,
+                .requests = std::move(configs),
+                .srv = octx.rctx.server().local(),
+                .mgr = mgr,
+                .as = _as.local(),
+              });
+          });
+
+        fill_fetch_responses(
+          octx,
+          std::move(results.read_results),
+          fetch.responses,
+          fetch.start_time);
+
+        // Remove errored fetches from fetch.requests and fetch.responses.
+        if (!results.errored_partitions.empty()) {
+            std::sort(
+              results.errored_partitions.begin(),
+              results.errored_partitions.end());
+            size_t shift = 0;
+            for (auto i : results.errored_partitions) {
+                fetch.requests.erase(fetch.requests.begin() + (i - shift));
+                fetch.responses.erase(fetch.responses.begin() + (i - shift));
+                shift += 1;
+            }
+        }
+
+        _last_result_size[shard] = results.total_size;
+        _completed_shard_fetches.push_back(std::move(fetch));
+        _has_data.signal();
+    }
+
+    std::optional<ss::future<>> register_hwm_wait(
+      shard_local_fetch_context& ctx,
+      const model::ktp& ktp,
+      model::offset offset,
+      ss::abort_source& as) {
+        auto part = ctx.mgr.get(ktp);
+        // If the partition can't be found then it's since been moved
+        if (!part) {
+            return {};
+        }
+
+        auto consensus = part->raft();
+        if (!consensus) {
+            return {};
+        }
+
+        return {ignore_exceptions(consensus->visible_offset_monitor().wait(
```

Comment:
Just a note, this is a great idea, however initially the `offset_monitor` probably was only meant to serve a couple of entities interested on propagation of the hwm, now the number of interested entities will exponentially grow. Looks like the underlying store is an absl btree multimap, seems ok for now but something to keep an eye on.

Reply:
Good point, none of the DT tests have failed due to an oversized allocation from it. However, I can see it potentially causing issues in the future. It really depends on how many consumers of a particular partition there is. Most workloads will likely only have 1 or 2 consumers per partition, so at most each partition offset monitor will see 1 or 2 new entries. However, you never know and a workload may come around that has 2,000+ consumers fetching from the same partition lol.